### DISCUSSION TITLE
Webdav: Could not connect, please try again

### DISCUSSION THREAD

**Discussion by ibigbug** [photoprism/photoprism#1692] [Q&A] [answered]
I'm connecting to Webdav Server running on my Synology to use the sync featuer.

But I'm getting error

> Could not connect, please try again

I can use other client to connect to the webdav server on Synology.

The log only said "could not connect" with Debug level enabled.

How can I debug this?

**Comment by lastzero** [comment]
You're trying to access your Synology NAS from PhotoPrism via WebDAV? HTTP or HTTPS? What specific WebDAV server & version (or client) is this?

**Comment by ibigbug** [comment]
Thanks for the reply.

> You're trying to access your Synology NAS from PhotoPrism via WebDAV? 

Yes


> HTTP or HTTPS? 

Tried both.


> What specific WebDAV server & version (or client) is this?

I actually don't know how to check this. I only installed the Webdav Server extension from Synology Store. do you know how to check that?

**Comment by lastzero** [comment]
Can you connect using other clients, for example your operating system?

**Comment by ibigbug** [comment]
Yes like I mentioned in the original post, I can use video player on my tablet to browse the files

**Comment by lastzero** [comment]
What video player specifically? What settings does it use?

There are many reasons a connection may fail:
- WebDAV server not reachable from your PhotoPrism instance (wrong URL, port, network, firewall,...)
- WebDAV server has an invalid HTTPS certificate
- WebDAV sever doesn't fully comply with standards (like when using PHP with nginx)
- Username or password are wrong
- Filesystem permission issues on the server
- ...

**Comment by ibigbug** [comment]
> What video player specifically? What settings does it use?

I used nPlayer

> There are many reasons a connection may fail:

Any way to print out a detailed error message so that I know where to start from rather than just saying "could not connect"

I enabled PhotoPrism debug log level but didn't see too much useful info from the loggings.

**Comment by lastzero** [comment]
PhotoPrism actually tests different paths and methods to figure out what works, so it wouldn't be one error but a potentially long list. For WebDAV, it tests the PROPFIND method on a number of common endpoints:

https://raw.githubusercontent.com/photoprism/photoprism/develop/internal/remote/heuristic.go

The response isn't even validated for this, so it shouldn't fail because of minor implementation details. Any 2xx response should do. My personal guess is that HTTP is disabled and the HTTPS certificate is invalid.

You may use curl - ideally from within the Docker container - for low level debugging as shown in these blog posts:

- https://www.qed42.com/blog/using-curl-commands-webdav
- https://code.blogs.iiidefix.net/posts/webdav-with-curl/

**Comment by lastzero** [comment]
Updated our FAQs: https://docs.photoprism.org/user-guide/faq/#why-do-i-get-an-error-when-trying-to-add-a-remote-server-for-syncing

**Comment by ibigbug** [comment]
very interesting.

I can use curl to auth but couldn't get anything on http://HOST:PORT/webdav/ by running:

```
$ curl -X GET  http://dsm:5005/webdav/ --basic --user 'user:pass'
```

HTTP is enabled.

it's just saying 404.

however I'm indeed able to use nPlayer to connect to the Webdav with no issues.

very interesting.

**Comment by ibigbug** [comment]
just btw, the `PROPFIND` is not implemented or disabled on the Synology NAS Webdav Server:

```
-> % curl -X PROPFIND -H "Depth: 1" -u user:pass http://dsm:5005/webdav/
<!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML 2.0//EN">
<html><head>
<title>405 Method Not Allowed</title>
</head><body>
<h1>Method Not Allowed</h1>
<p>The requested method PROPFIND is not allowed for this URL.</p>
</body></html>
```

**Comment by lastzero** [comment]
All DAV compliant resources must support the PROPFIND method 👇

http://www.webdav.org/specs/rfc2518.html#METHOD_PROPFIND

**Comment by ibigbug** [comment]
I think I found it.

The server is running ok, I just needed to remove the `/webdav/` and the testing with cURL was no issue.

The problem in this issue is in the Heuristic it's assuming the webdav server has "annonymous" access enabled by running `HttpOk`  check, https://github.com/photoprism/photoprism/blob/19e9c7560ebbdd7d8b77773a509adead82865d70/internal/remote/heuristic.go#L57

which will fail in for the server who denies anounymous access at all, which is the case of my NAS setup.

@lastzero am I making sense?

**Comment by ibigbug** [comment]
ping

**Comment by lastzero** [comment]
Access shouldn't be anonymous, looking at the request:

```
if user != "" || pass != "" {
		u.User = url.UserPassword(result.AccUser, result.AccPass)
	}
```

Note I'm currently working on facial recognition, so I don't have much time taking care of other issues.

**Comment by ibigbug** [comment]
I’m talking about the `HttpOk` method, which is used to test the link rather than the actual accessing.

If you don’t have time I can quickly draft a PR to fix this.

**Comment by lastzero** [comment]
The function is trying to access a URL containing the user and password from what I see, see the code I posted above. Are you sure the request is anonymous? You'd expect this to fail, I agree.

**Comment by ibigbug** [comment]
I see what you mean.

Thought the u.User is some form of user:pass@URI described in https://www.rfc-editor.org/rfc/rfc2396.html

But WebDAV is expecting the “Authentication” header,  http://www.webdav.org/specs/rfc2518.html#n-authentication-of-clients

Might be the cause?

**Comment by lastzero** [comment]
Should be sent as Authentication header, not aware of any other way. Check the http client implementation to be sure.

**Comment by lastzero** [comment]
Checking back later this week, it's past 1am over here.

**Comment by ibigbug** [comment]
Checked, you are right https://cs.opensource.google/go/go/+/refs/tags/go1.17:src/net/http/client.go;l=244;drc=refs%2Ftags%2Fgo1.17

Will come back later

**Comment by averymd** [comment, marked as answer]
I also run a Synology NAS and Photoprism would not connect until I checked the box for "Enable DavDepthInfinity". @ibigbug, have you tried that setting?

**Comment by graciousgrey** [comment]
@ibigbug is your problem solved?

**Comment by Bananas-Are-Yellow** [comment]
Same problem here.

Yes, the problem is solved if you check _Enable DavDepthInfinity_ in the Synology WebDAV Server settings. But I would argue this shouldn't be necessary. That setting is disabled by default because `Depth: infinity` could be expensive.

The problem arises because `Depth: infinity` is the default if the depth is not specified.

http://www.webdav.org/specs/rfc2518.html#METHOD_PROPFIND

> A client may submit a Depth header with a value of "0", "1", or "infinity" with a PROPFIND on a collection resource with internal member URIs. DAV compliant servers MUST support the "0", "1" and "infinity" behaviors. By default, the PROPFIND method without a Depth header MUST act as if a "Depth: infinity" header was included.

The moral of the story is that you should _always_ specify the depth (usually 0 or 1) in every use of PROPFIND, because the default of `Depth: infinity` is usually not what you want and could be expensive. I imagine `Depth: 0` is all that's needed for simple tests to confirm the connection.
